src/storage/copier/node_copy_executor.cpp

```diff
@@ -0,0 +1,105 @@
+}
+
+void NodeCopyExecutor::populateColumnsAndLists(processor::ExecutionContext* executionContext) {
+    auto primaryKey = reinterpret_cast<NodeTableSchema*>(tableSchema)->getPrimaryKey();
+    switch (primaryKey.dataType.typeID) {
+    case INT64: {
+        populateColumns<int64_t>(executionContext);
+    } break;
+    case STRING: {
+        populateColumns<ku_string_t>(executionContext);
+    } break;
+    default: {
+        throw CopyException(StringUtils::string_format("Unsupported data type {} for the ID index.",
+            Types::dataTypeToString(primaryKey.dataType)));
+    }
+    }
+}
+
+template<typename T>
+void NodeCopyExecutor::populateColumns(processor::ExecutionContext* executionContext) {
+    logger->info("Populating properties");
+    auto primaryKey = reinterpret_cast<NodeTableSchema*>(tableSchema)->getPrimaryKey();
+    auto pkIndex = std::make_unique<HashIndexBuilder<T>>(
+        StorageUtils::getNodeIndexFName(
+            this->outputDirectory, tableSchema->tableID, common::DBFileType::WAL_VERSION),
+        primaryKey.dataType);
+    pkIndex->bulkReserve(numRows);
+
+    std::vector<std::shared_ptr<common::Task>> tasks;
+    switch (copyDescription.fileType) {
+    case common::CopyDescription::FileType::CSV: {
+        std::shared_ptr<NodeCopySharedState> sharedState =
+            std::make_shared<CSVNodeCopySharedState>(copyDescription.filePaths, fileBlockInfos,
+                copyDescription.csvReaderConfig.get(), tableSchema);
+        auto nodeCopier = std::make_unique<CSVNodeCopier<T>>(std::move(sharedState), pkIndex.get(),
+            copyDescription, columns, propertyIDToColumnIDMap[primaryKey.propertyID]);
+        tasks.push_back(std::make_shared<NodeCopyTask<T>>(std::move(nodeCopier), executionContext));
+    } break;
+    case common::CopyDescription::FileType::PARQUET: {
+        std::shared_ptr<NodeCopySharedState> sharedState =
+            std::make_shared<NodeCopySharedState>(copyDescription.filePaths, fileBlockInfos);
+        auto nodeCopier =
+            std::make_unique<ParquetNodeCopier<T>>(std::move(sharedState), pkIndex.get(),
+                copyDescription, columns, propertyIDToColumnIDMap[primaryKey.propertyID]);
+        tasks.push_back(std::make_shared<NodeCopyTask<T>>(std::move(nodeCopier), executionContext));
+    } break;
+    case common::CopyDescription::FileType::NPY: {
+        for (auto i = 0u; i < copyDescription.filePaths.size(); i++) {
+            std::vector<std::string> filePaths(1);
+            filePaths[0] = copyDescription.filePaths[i];
+            std::shared_ptr<NodeCopySharedState> sharedState =
+                std::make_shared<NodeCopySharedState>(filePaths, fileBlockInfos);
+            auto nodeCopier = std::make_unique<NPYNodeCopier<T>>(std::move(sharedState),
+                pkIndex.get(), copyDescription, columns, i,
+                i == propertyIDToColumnIDMap[primaryKey.propertyID] ? 0 : INVALID_COLUMN_ID);
+            tasks.push_back(
```

Comment:
you can probably pull task.push_back out of the case switch

Reply:
Not really, npy is specially handled right now, as each file is assigned a task.